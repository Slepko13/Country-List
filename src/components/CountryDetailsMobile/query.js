import {gql} from "@apollo/client";

export const COUNTRY = gql`
    query GetCountry($name: String)  {
        Country(name: $name) {
            name
            population
            capital
            officialLanguages {
                name
            }
            subregion {
                name
                region {
                    name
                }
            }
            currencies {
                name
                symbol
            }
            flag {
                svgFile
            }
            callingCodes {
                name
            }
            timezones {
                name
            }
        }
    }
`;
