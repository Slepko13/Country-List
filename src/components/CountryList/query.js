import {gql} from "@apollo/client";

export const COUNTRIES = gql`
    {
        Country(orderBy: name_asc) {
            name
            capital
            subregion {
                region {
                    name
                }
            }
            flag{
                svgFile
            }

        }
    }
`;
