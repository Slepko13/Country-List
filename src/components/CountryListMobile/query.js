import { gql } from "@apollo/client";

export const COUNTRIES = gql`
    {
        countries(filter:{}) {
            name
            capital
            code
            continent {
            name
            }
        }
    }
`;
