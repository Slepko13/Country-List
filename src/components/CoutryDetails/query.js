import {gql} from "@apollo/client";

export const COUNTRY = gql`
    query GetCountry($name: String)  {
        Country
        (name: $name)
        {
            name
            population
            capital
            officialLanguages {
                name
            }
            subregion {
                region {
                    name
                }
            }
            currencies {
                name
                symbol
            }
            flag {
                svgFile
            }
            callingCodes {
                name
            }
            timezones {
                name
            }
        }
    }
`;
