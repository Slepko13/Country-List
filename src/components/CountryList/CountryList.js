import React, {useState} from 'react';
import './CountryList.scss';

import { useQuery} from '@apollo/client';
import {COUTRIES} from "./query";





function CountryList({setId}) {
    const { loading, error, data } = useQuery(COUTRIES);

    if (loading) return <div className="CountryList "><p className="loading">Loading...</p></div>;
    if (error) return <div className="CountryList"><p className="error">Error :(</p></div>;

    const renderList = (list) => (

    <div className="CountryList ">
      <div className="list">

          {list.map(({name, capital , flag : {svgFile : flag}, subregion }, index ) =>
              {
                  return        (
                      <div
                          className="list__item"
                          key={name}
                          onClick={(e) => { setId(name) }}

                      >
                          <div className="item__flag">
                              <img className="flag__image"  src={flag} alt="flag"/>
                          </div>
                          <div className="item__info border">
                              <div className="info__country">
                                  <div className="info__title">Country:</div>
                                  <div className="info__data">{name}</div>
                                  {/*<div className="full__data">{name}</div>*/}

                              </div>
                              {capital === "Kiev" ?
                                  <div className="info__capital">
                                      <div className="info__title">Capital:</div>
                                      <div className="info__data">Kyiv</div>
                                  </div>:
                                  capital==="" ?
                                      null :
                                      <div className="info__capital">
                                          <div className="info__title">Capital:</div>
                                          <div className="info__data">{capital}</div>
                                      </div>
                              }
                              { subregion ?
                                  <div className="info__region">
                                      <div className="info__title">Region:</div>
                                      <div className="info__data">{subregion.region.name}</div>
                                  </div> :

                                  null}
                          </div>
                      </div>
                  )
              }

          )}
      </div>
        </div>
    )
    return renderList(data.Country);
}

export default CountryList;