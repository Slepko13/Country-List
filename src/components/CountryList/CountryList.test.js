import React from "react";
import {MockedProvider} from '@apollo/client/testing';
import {act, render, fireEvent, cleanup} from '@testing-library/react';

import {COUNTRIES} from "./query";
import CountryList from "./CountryList";


const errorMock = [{
    request: {
        query: COUNTRIES
    },
    error: new Error('aw shucks'),
}];

const dataMocks = [
    {
        request: {
            query: COUNTRIES
        },
        result: {
            data: {
                Country: [{
                    name: 'France',
                    capital: 'Paris',
                    subregion: {region: {name: 'Europe'}},
                    flag: {svgFile: "flag"}
                }]
            }
        }
    }
];
const dataMocksKiev = [
    {
        request: {
            query: COUNTRIES
        },
        result: {
            data: {
                Country: [{
                    name: 'Ukraine',
                    capital: 'Kiev',
                    subregion: {region: {name: 'Europe'}},
                    flag: {svgFile: "flag"}
                }]
            }
        }
    }
];

const dataMocksEmptyCapital = [
    {
        request: {
            query: COUNTRIES
        },
        result: {
            data: {
                Country: [{
                    name: 'Ukraine',
                    capital: '',
                    subregion: {region: {name: 'Europe'}},
                    flag: {svgFile: "flag"}
                }]
            }
        }
    }
];
const dataMocksEmptySubregion = [
    {
        request: {
            query: COUNTRIES
        },
        result: {
            data: {Country: [{name: 'France', capital: 'Paris', subregion: null, flag: {svgFile: "flag"}}]}
        }
    }
];
const dataMocksEmpty = [
    {
        request: {
            query: COUNTRIES
        },
        result: {
            data: {Country: []}
        }
    }
];

async function wait(ms = 0) {
    await act(() => {
        return new Promise(resolve => {
            setTimeout(resolve, ms);
        });
    });
}

afterEach(cleanup);

describe('CountryList', () => {
    it('renders', async () => {
        const {asFragment} = render(
            <MockedProvider addTypename={false} mocks={dataMocks}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders error', async () => {
        const {container} = render(
            <MockedProvider addTypename={false} mocks={errorMock}>
                <CountryList/>
            </MockedProvider>
        );

        await wait();

        expect(container.textContent).toMatch('Sorry,  you got an error: some error');
    });

    it('renders correct texts in card ', async () => {
        const {container, getByTestId} = render(
            <MockedProvider addTypename={false} mocks={dataMocks}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(getByTestId("France")).toHaveTextContent("Europe");
        expect(getByTestId("France")).toHaveTextContent("Paris");
        expect(getByTestId("France")).toHaveClass("list__item");
        expect(container.textContent).toMatch('France');
    });

    it('renders correct texts in card with Kyiv capital', async () => {
        const {container, getByTestId} = render(
            <MockedProvider addTypename={false} mocks={dataMocksKiev}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(getByTestId("Ukraine")).toHaveTextContent("Europe");
        expect(getByTestId("Ukraine")).toHaveClass("list__item");
        expect(container.textContent).toMatch('Ukraine');
        expect(container.textContent).toMatch('Kyiv');
    });

    it('renders without capital', async () => {
        const {queryByTestId} = render(
            <MockedProvider addTypename={false} mocks={dataMocksEmptyCapital}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(queryByTestId('customCapital')).toBeFalsy();
        expect(queryByTestId('KyivCapital')).toBeFalsy();
    });

    it('renders without subregion', async () => {
        const {queryByTestId} = render(
            <MockedProvider addTypename={false} mocks={dataMocksEmptySubregion}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(queryByTestId('subregion')).toBeFalsy();
    });

    it('renders without data', async () => {
        const {getByText} = render(
            <MockedProvider addTypename={false} mocks={dataMocksEmpty}>
                <CountryList setId={() => {
                }}/>
            </MockedProvider>
        );
        await wait();
        expect(getByText('Empty data')).toBeTruthy();
    });

    it('setId should be called', async () => {
        const setIdSpy = jest.fn();
        const {getByTestId} = render(
            <MockedProvider addTypename={false} mocks={dataMocks}>
                <CountryList setId={setIdSpy}/>
            </MockedProvider>
        );
        await wait();
        fireEvent.click(getByTestId("France"));
        expect(setIdSpy).toHaveBeenCalledWith("France");
    });
});