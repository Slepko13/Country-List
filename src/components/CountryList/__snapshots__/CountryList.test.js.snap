// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList renders 1`] = `
<DocumentFragment>
  <div
    class="CountryList "
  >
    <div
      class="list"
    >
      <div
        class="list__item"
        data-testid="France"
      >
        <div
          class="item__flag"
        >
          <img
            alt="flag"
            class="flag__image"
            src="flag"
          />
        </div>
        <div
          class="item__info"
        >
          <div
            class="info__country"
          >
            <div
              class="info__title"
            >
              Country:
            </div>
            <div
              class="info__data"
            >
              France
            </div>
          </div>
          <div
            class="info__capital"
            data-testid="customCapital"
          >
            <div
              class="info__title"
            >
              Capital:
            </div>
            <div
              class="info__data"
            >
              Paris
            </div>
          </div>
          <div
            class="info__region"
            data-testid="subregion"
          >
            <div
              class="info__title"
            >
              Region:
            </div>
            <div
              class="info__data"
            >
              Europe
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
